File: app/src/test/java/com/github/goldy1992/mp3player/service/library/content/retriever/SongsFromFolderRetrieverTest.java
package com.github.goldy1992.mp3player.service.library.content.retriever;

import android.support.v4.media.MediaBrowserCompat;

import com.github.goldy1992.mp3player.commons.MediaItemBuilder;
import com.github.goldy1992.mp3player.service.library.content.parser.SongResultsParser;
import com.github.goldy1992.mp3player.service.library.content.request.ContentRequest;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;

import java.util.List;

import static android.os.Looper.getMainLooper;
import static com.github.goldy1992.mp3player.commons.MediaItemType.SONG;
import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.when;
import static org.robolectric.Shadows.shadowOf;

@RunWith(RobolectricTestRunner.class)
public class SongsFromFolderRetrieverTest extends ContentResolverRetrieverTestBase<SongsFromFolderRetriever> {

    @Mock
    SongResultsParser resultsParser;

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);
        this.retriever = spy(new SongsFromFolderRetriever(contentResolver, resultsParser, null, handler));
    }

    @Test
    public void testGetChildren() {
        this.contentRequest = new ContentRequest("x", "y", "z");
        final String id = "xyz";
        final String title = "title";
        MediaBrowserCompat.MediaItem mediaItem = new MediaItemBuilder(id)
                .setTitle(title)
                .build();
        expectedResult.add(mediaItem);
        when(contentResolver.query(any(), any(), any(), any(), eq(null))).thenReturn(cursor);
        when(resultsParser.create(cursor, contentRequest.getMediaIdPrefix())).thenReturn(expectedResult);


        List<MediaBrowserCompat.MediaItem> result = retriever.getChildren(contentRequest);
        // call remaining looper messages
        shadowOf(getMainLooper()).idle();

        // assert results are the expected ones
        assertEquals(expectedResult, result);

    }
    @Test
    @Override
    public void testGetMediaType() {
        assertEquals(SONG, retriever.getType());
    }
}

File: service/src/testFullDebug/java/com/github/goldy1992/mp3player/service/library/content/retriever/SongsFromFolderRetrieverTest.kt
package com.github.goldy1992.mp3player.service.library.content.retriever

import android.os.Looper
import com.github.goldy1992.mp3player.commons.MediaItemBuilder
import com.github.goldy1992.mp3player.commons.MediaItemType
import com.github.goldy1992.mp3player.service.library.content.parser.SongResultsParser
import com.github.goldy1992.mp3player.service.library.content.retrievers.SongsFromFolderRetriever
import com.github.goldy1992.mp3player.service.library.data.search.SongDao
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.kotlin.any
import org.mockito.kotlin.eq
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import org.robolectric.RobolectricTestRunner
import org.robolectric.Shadows

@RunWith(RobolectricTestRunner::class)
class SongsFromFolderRetrieverTest : ContentResolverRetrieverTestBase<SongsFromFolderRetriever?>() {

    var resultsParser: SongResultsParser = mock<SongResultsParser>()

    var songDao: SongDao = mock<SongDao>()

    @Before
    fun setup() {
        retriever = SongsFromFolderRetriever(contentResolver, resultsParser, null)
    }

    @Test
    fun testGetChildren() {
        val id = "xyz"
        val title = "title"
        val mediaItem = MediaItemBuilder(id)
                .setTitle(title)
                .build()
        expectedResult.add(mediaItem)
        whenever(contentResolver.query(any(), any(), any(), any(), eq(null))).thenReturn(cursor)
        whenever(resultsParser.create(cursor)).thenReturn(expectedResult)
        val result = retriever!!.getChildren(id)
        // call remaining looper messages
        Shadows.shadowOf(Looper.getMainLooper()).idle()
        // assert results are the expected ones
        Assert.assertEquals(expectedResult, result)
    }

    @Test
    override fun testGetMediaType() {
        Assert.assertEquals(MediaItemType.SONG, retriever!!.type)
    }
}